File: frontend/src/services/task-service.ts
import { request } from '@/lib/api-client';
import { PaginatedResponse } from '@/types/common';
import { Task, TaskCreate, TaskUpdate, TaskStatus, TaskPriority } from '@/types/task';

export const taskService = {
  getTasks: (params?: {
    status?: TaskStatus;
    priority?: TaskPriority;
    q?: string;
    page?: number;
    page_size?: number;
    sort_by?: string;
    sort_dir?: string;
    token?: string;
  }) => {
    const { token, ...queryParams } = params || {};
    return request<PaginatedResponse<Task>>('/tasks/', {
      token,
      params: queryParams,
    });
  },

  getTask: (id: string) => request<Task>(`/tasks/${id}`),

  createTask: (data: TaskCreate) =>
    request<Task>('/tasks/', {
      method: 'POST',
      body: JSON.stringify(data),
    }),

  updateTask: (id: string, data: TaskUpdate) =>
    request<Task>(`/tasks/${id}`, {
      method: 'PUT',
      body: JSON.stringify(data),
    }),

  updateTaskStatus: (id: string, status: TaskStatus) =>
    request<Task>(`/tasks/${id}/status`, {
      method: 'PATCH',
      body: JSON.stringify({ status }),
    }),

  deleteTask: (id: string) =>
    request<void>(`/tasks/${id}`, {
      method: 'DELETE',
    }),
};

File: backend/app/services/task_service.py
import uuid
from typing import Optional
from ..repositories.task_repository import TaskRepository
from ..models.task import Task, TaskStatus
from ..schemas.task import TaskCreate, TaskUpdate

from .base_service import BaseService

class TaskService(BaseService[Task, TaskRepository]):
    def __init__(self, task_repo: TaskRepository):
        super().__init__(task_repo, "task")

    def get_tasks(
        self,
        user_id: uuid.UUID,
        page: int,
        page_size: int,
        status: Optional[str] = None,
        priority: Optional[str] = None,
        q: Optional[str] = None,
        sort_by: str = "created_at",
        sort_dir: str = "desc",
    ):
        return self.repository.get_all(
            user_id=user_id,
            page=page,
            page_size=page_size,
            status=status,
            priority=priority,
            q=q,
            sort_by=sort_by,
            sort_dir=sort_dir,
        )

    def get_task(self, user_id: uuid.UUID, task_id: uuid.UUID) -> Task:
        return self.get_or_404(user_id, task_id)

    def create_task(self, user_id: uuid.UUID, task_data: TaskCreate) -> Task:
        return self.create_with_commit(user_id, task_data)

    def update_task(
        self, user_id: uuid.UUID, task_id: uuid.UUID, task_data: TaskUpdate
    ) -> Task:
        return self.update_with_commit(user_id, task_id, task_data)

    def delete_task(self, user_id: uuid.UUID, task_id: uuid.UUID) -> bool:
        return self.delete_with_commit(user_id, task_id)

    def update_task_status(
        self, user_id: uuid.UUID, task_id: uuid.UUID, new_status: TaskStatus
    ) -> Task:
        task = self.get_or_404(user_id, task_id)
        task.status = new_status
        return self.repository.commit_and_refresh(task)
